### ISSUE TITLE
copy elision when LHS and RHS have different types

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#17089]
We have copy elision that applies when the RHS initializing a variable is not used again. However the compiler does not currently allow copy elision when the types are different.

One test that relies on this behavior is  test/studies/shootout/nbody/sidelnik/nbody_orig_1.chpl.

<details>

After PR #16802, array literals are created with `chpl__buildArrayExpr` using the `in` intent. The result is that the code
    
      var bodies: [0..#NBODIES] borrowed Planet = [ new Planet(...), ... ]
    
creates a RHS temporary array of owned - let's call it `tmpArray`. `tmpArray` is destroyed at the end of the block (and the temporaries for the individual `new` call results are moot at this point since they are `nil` from having been transferred out of).
    
This was working as of PR #16802 because the array `bodies` array was created by copy-initializing from `tmpArray`. However, if this is changed to move-initialization, we get a use-after-free (because the `borrowed` array refers to something that is destroyed).

</details>

Should copy elision apply when the LHS and RHS types in the initialization differ?

Arguments in favor:
 * It makes the copy elision more consistent if it is type independent

Arguments against:
 * When the types differ, it's more of a cast operation, and "stealing" the memory from the RHS will probably not be possible
 
 
The language specification does not say anything about a type constraint for copy elision. So, if the answer is "yes", the implementation needs adjustment, and if the answer is "no", the spec needs adjustment.

**Comment by bradcray** [comment]
I'm not sure I have a strong opinion about the overall question at this time, though my instinct is that we shouldn't cast-and-steal in the general / value case.  Classes of course _may_ be somewhat different given (a) relationships between parent and child classes where casting isn't involved, (b) nilable vs. non-nilable classes using the same storage, or (c) borrowed and unmanaged classes using the same storage.

Considering the motivating example from the OP:

```chapel
var bodies: [0..#NBODIES] borrowed Planet = [ new Planet(...), ... ]
```

it seems to me that it'd be preferable if this code were written such that `bodies` was defined as an array of `owned Planet` as I don't think the code should rely on the array literal retaining the ownership of the anonymous objects in cases like this.  But I don't think I'd expect copy elision to fire here given that owned classes and borrowed classes have different storage.

**Comment by mppf** [comment]
Right, that nbody example is pretty strange (and something I added to the test when adding owned/shared/etc).

I think the argument for allowing copy elision with different RHS sides is probably something more along the lines of - say you have two lists/arrays/etc with a difference in type that does not affect storage. For example maybe they have a different value for a `param locking`. In that case, the LHS could steal from the RHS, but it would only make sense for certain types. In that event - should the compiler try to call some sort of cast-move function if it exists?

I suppose either way that we shouldn't switch to a move without calling some sort of special operator/method. For arrays it was already calling a module code function for the move (because of runtime types).

So, I think probably the answer is two fold:
 1. The immediate answer is "no" and we need to update the spec
 2. If in the future if we decide to support such functionality for user-defined types we will have to define a move-cast operation that the compiler can call, when the type author opts in to it.
 
 I am not sure what that means for the number of copies occurring for promoted array assignment (var A:[1..10] R = myRecord) today. I'm inclined to leave out the copy and leave optimizing it away until we have the general feature described in (2), but that does arguably introduce a very minor language instability. (But... in a promoted case like this, the number of copies also depends on the number of tasks created today, so it's not really something one could reasonably rely on anyway).
